File: rtiow/Nim/vec3.nim
import math

type Vec3* = object
    x*, y*, z*: float32

{.push inline.}

func vec3*(x, y, z: float32): Vec3 =
    Vec3(x: x, y: y, z: z)

func squared_length*(v: Vec3): float32 =
    v.x * v.x + v.y * v.y + v.z * v.z

func length*(v: Vec3): float32 =
    sqrt(v.squared_length())

func dot*(v1, v2: Vec3): float32 =
    v1.x * v2.x + v1.y * v2.y + v1.z * v2.z

func `/`*(v: Vec3, t: float32): Vec3 =
    result.x = v.x / t
    result.y = v.y / t
    result.z = v.z / t

func `*`*(t: float32, v: Vec3): Vec3 =
    result.x = t * v.x
    result.y = t * v.y
    result.z = t * v.z

func `+`*(v1: Vec3, v2: Vec3): Vec3 =
    result.x = v1.x + v2.x
    result.y = v1.y + v2.y
    result.z = v1.z + v2.z

func `-`*(v1: Vec3, v2: Vec3): Vec3 =
    result.x = v1.x - v2.x
    result.y = v1.y - v2.y
    result.z = v1.z - v2.z

func `*`*(v1: Vec3, v2: Vec3): Vec3 =
    result.x = v1.x * v2.x
    result.y = v1.y * v2.y
    result.z = v1.z * v2.z

func unit_vector*(v: Vec3): Vec3 =
    v / v.length()

func vreflect*(v: Vec3, n: Vec3): Vec3 =
    v - 2f * dot(v, n) * n

{.pop.}

File: rtiow/JS/vec3.js
function vsub(v1, v2) {
    v1[0] -= v2[0]; v1[1] -= v2[1]; v1[2] -= v2[2];
    // return [v1[0] - v2[0], v1[1] - v2[1], v1[2] - v2[2]];
    // return [parseFloat(v1[0]) - v2[0], parseFloat(v1[1]) - v2[1], parseFloat(v1[2]) - v2[2]];
    // return [parseFloat(v1[0] - v2[0]), parseFloat(v1[1] - v2[1]), parseFloat(v1[2] - v2[2])];
    // return new Float32Array([v1[0] - v2[0], v1[1] - v2[1], v1[2] - v2[2]]);
}
function vadd(v1, v2) {
    v1[0] += v2[0]; v1[1] += v2[1]; v1[2] += v2[2];
    // return [v1[0] + v2[0], v1[1] + v2[1], v1[2] + v2[2]];
    // return [parseFloat(v1[0]) + v2[0], parseFloat(v1[1]) + v2[1], parseFloat(v1[2]) + v2[2]];
    // return new Float32Array([v1[0] + v2[0], v1[1] + v2[1], v1[2] + v2[2]]);
}
function vmul(f, v) {
    // return [f * v[0], f * v[1], f * v[2]];
    v[0] *= f; v[1] *= f; v[2] *= f;
    // return [parseFloat(f) * v[0], parseFloat(f) * v[1], parseFloat(f) * v[2]];
    // return new Float32Array([f * v[0], f * v[1], f * v[2]]);
}
function vdiv(v, f) {
    // const inv = 1. / f;
    // v[0] *= f, v[1] / f, v[2] / f];
    vmul(1. / f, v);
    // return [v[0] / f, v[1] / f, v[2] / f];
    // return [inv * v[0], inv * v[1], inv * v[2]];
    // return [parseFloat(v[0]) / f, parseFloat(v[1]) / f, parseFloat(v[2]) / f];
    // return new Float32Array([v[0] / f, v[1] / f, v[2] / f]);
}
function dot(v1, v2) {
    return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
    // return parseFloat(v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2]);
    // return parseFloat(v1[0]) * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
}
function vlensq(v) {
    // return dot(v, v);
    return v[0] * v[0] + v[1] * v[1] + v[2] * v[2];
}
function vlen(v) {
    return Math.sqrt(vlensq(v));
}
function unit_vector(v) {
    //    return vdiv(v, vlen(v));
    vdiv(v, vlen(v));
}
